src/dialogs/alignmentarray.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2022 Meltytech, LLC
+ *
+ * Author: André Caldas de Souza <andrecaldas@unb.br>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "alignmentarray.h"
+
+#include <QDebug>
+#include <QMutexLocker>
+
+#include <algorithm>
+#include <cstring>
+#include <cmath>
+#include <iostream>
+#include <numeric>
+
+// FFTW plan functions are not threadsafe
+static QMutex s_fftwPlanningMutex;
+
+AlignmentArray::AlignmentArray()
+    : m_autocorrelationMax(std::numeric_limits<double>::min())
+    , m_isTransformed(false)
+{
+}
+
+AlignmentArray::AlignmentArray(size_t minimum_size)
+    : AlignmentArray()
+{
+    init(minimum_size);
+}
+
+AlignmentArray::~AlignmentArray()
+{
+    QMutexLocker locker(&s_fftwPlanningMutex);
+    fftw_free(reinterpret_cast<fftw_complex*>(m_forwardBuf));
+    fftw_destroy_plan(m_forwardPlan);
+}
+
+void AlignmentArray::init(size_t minimumSize)
+{
+    m_minimumSize = minimumSize;
+    m_actualComplexSize = (minimumSize * 2) - 1;
+}
+
+void AlignmentArray::setValues(const std::vector<double>& values)
+{
+    m_values = values;
+}
+
+double AlignmentArray::calculateOffset(AlignmentArray &from, int* offset)
+{
+    // Create a destination for the correlation values
+    s_fftwPlanningMutex.lock();
+    fftw_complex* buf = fftw_alloc_complex(m_actualComplexSize);
+    std::complex<double>* correlationBuf = reinterpret_cast<std::complex<double>*>(buf);
+    fftw_plan correlationPlan = fftw_plan_dft_1d(m_actualComplexSize, buf, buf, FFTW_BACKWARD, FFTW_ESTIMATE);
+    std::fill(correlationBuf, correlationBuf + m_actualComplexSize, std::complex<double>(0));
+    s_fftwPlanningMutex.unlock();
+
+    // Ensure the two sequences are transformed
+    transform();
+    from.transform();
+
+    // Calculate the cross-correlation signal
+    for(size_t i = 0; i < m_actualComplexSize; ++i)
+    {
+        correlationBuf[i] = m_forwardBuf[i] * std::conj(from.m_forwardBuf[i]);
+    }
```

Comment:
The coding style in this file does not conform. I realize you did not write it, but I want it reformatted.

Reply:
done